### PR TITLE
file: unix: macos: Resize file chunks if needed

### PR DESCRIPTION
**Pull request by cosmo0920**
Currently, resizing operation is always kicked and resized for mmap-ed files.
However, on new_size < cf->alloc_size case, it will cause invalid status of mmap-ed files on macOS.

### CODE REVIEW

**Review by leonardo-albertovich — CHANGES_REQUESTED** [review]
This is a placeholder, please do not merge this PR until I approve it.

File: src/cio_file_unix.c
Comment (edsiper) [inline src/cio_file_unix.c:122]: isn't the code suggested the same as cio_file_native_unmap(cf); ?

Comment (cosmo0920) [reply src/cio_file_unix.c]: Indeed. I changed it in https://github.com/fluent/chunkio/pull/93/commits/1adcd3743b3d74438b2a3ead9cda1c4b17d9df15.

**Comment by cosmo0920** [comment]
Actually, we only need to add size comparison for new_size and cf->alloc_size on macOS.
For Linux, adding another clause to compare new_size and cf->alloc_size is too pessimistic.

**Comment by cosmo0920** [comment]
I re-evaluated this patch and I found that only adding comparison between new_size and cf->alloc_size is enough.

File: src/cio_file_unix.c
Comment (edsiper) [inline src/cio_file_unix.c:125]: if this means we cannot shrink a chunk "I think" we might have the following unhandled cases:

- A chunk is always created with extra size, when the Chunk is putted in a down state, the size needs to be adjusted to reduce the extra space added at the beginning.

please re-validate this

Comment (cosmo0920) [reply src/cio_file_unix.c:125]: Hmm..., got it. I'm re-evaluate this.

Comment (cosmo0920) [reply src/cio_file_unix.c:125]: As far as I observed, on macOS, I cannot run unmap/mmap without error less than page_size(16384).

Comment (cosmo0920) [reply src/cio_file_unix.c]: > * the size needs to be adjusted to reduce the extra space added at the beginning.

This could be caused memory corruption.

On the other hand, the following patch is also able to consider to handle mremap emulation on macOS:

```diff
diff --git a/src/cio_file.c b/src/cio_file.c
index 786a022..b4d5218 100644
--- a/src/cio_file.c
+++ b/src/cio_file.c
@@ -1130,6 +1130,17 @@ int cio_file_resize(struct cio_file *cf, size_t new_size)
     mapped_flag = cio_file_native_is_mapped(cf);
     mapped_size = cf->alloc_size;
 
+#ifdef __APPLE__
+    /* On macOS, when the size of new_size is less than pagesize,
+     * unmap/mmap operations will cause unexpected behavior such as
+     * memory corruptions.
+     * This could make diffucult to handle further operations.
+     * In this case, prohibit to shrink file size. */
+    if (cio_getpagesize() > new_size) {
+        return CIO_OK;
+    }
+#endif
+
 #ifdef _WIN32
     if (mapped_flag) {
         result = cio_file_native_unmap(cf);
```

However, this causes checksum corruptions.

Comment (cosmo0920) [reply src/cio_file_unix.c]: So, my thought is:

* Prevent to shrink less than page size (16384 bytes for macOS-arm64) on macOS.
* Call cio_file_native_resize due to accurate calculations for checksum

Comment (edsiper) [reply src/cio_file_unix.c]: maybe the re-map is not needed, because if it needs to be shrink the process can be:

- sync changes to disk
- unmap the memory region
- truncate the file to the desired size
- mmap() again if needed

Comment (cosmo0920) [reply src/cio_file_unix.c]: Yes. Intending to implement less than page size operation on macOS is dangerous.
For syncing contents of actual mmap-ed files, we just call msync to handle written characters in mmap-ed regions.

The root cause of this issue should be cause by less mmap-ed size than pagesize. Basically, the page size of macOS should be 4096 in Intel Mac, and should be 16384 on Apple Silicon. In macOS, there is no ability to shrink less than pagesize.
Other platforms can be used for mmap-ed files less than 4096/16384.

For Linux, we can use `mremap` to shrink mmap-ed region less than pagesize. This could be valid operation on Linux.
For Windows, we can use `CreateFileMappingA` and `MapViewOfFile` to map less than ordinary pagesize. This also could be valid operation on Windows.
Although, in macOS, there is no way to handle less than page size. 
It cause always 0xfffffffff(-1) on start to read mmap-ed contents....

**Comment by cosmo0920** [comment]
This issue could be caused by TOCTOU. This approach should conceal what is the real culprit.
It should be handled in fluent-bit side.
Thank you for your investigation, @leonardo-albertovich !

**Pull request closed without merging**
